pios_video: change suspected bugs

mlyle opened the pull request:
I am unable to fully verify this-- or test this at all-- this moment, but I've noticed several suspect things in pios_video in code audit, that could do really bad things if hsync is toggling too much (for no good reason)

Also, it's possible to filter the video hsync, so let's try enabling it.

Fixes #1588

mlyle committed:
pios_video: eliminate dependency on TIM4 line counter

mlyle committed:
pios_video: remove line_counter conf field

mlyle committed:
pios_video: slight tweaks

- add a mode hysteresis value so we're not willing to change modes based
  on one odd frame
- adjust graphics_line_start based on slight behavior change

mlyle committed:
playuavosd: adjust for line counter change

mlyle committed:
pios_video: fix dumb bug with mode hysteresis

mlyle committed:
seppuku: enable PIOS_VIDEO_INPUT_FILTER

mlyle commented:
jenkins, build artifacts please

mlyle commented:
Main commit comment:

1) ~~we re-enable the line counter for unclear reasons during the frame.
We certainly don't **use** the line counter during a frame.~~
2) It looks like we don't clear the line counter interrupt flag in all
cases.  Surely this is bad if we get a excessively high number of lines?
3) Add a PIOS_VIDEO_INPUT_FILTER thingy that lets us require a minimum
hsync width to trigger.
commit a07ceb6372dd6ec9184c2e26ff35ff13fc774c66

mlyle committed:
pios_video: improvement to sync handling

1. It looks like we don't clear the line counter interrupt flag in all
cases. Surely this is bad if we get a excessively high number of lines?

2. Add a PIOS_VIDEO_INPUT_FILTER thingy that lets us require a minimum
hsync width to trigger.

tracernz commented on flight/PiOS/STM32F4xx/pios_video.c:
```diff
@@ -367,15 +367,34 @@ void PIOS_Video_Init(const struct pios_video_cfg *cfg)
-	tmpccer |= (uint16_t)(TIM_CCER_CC1P);
-	cfg->hsync_capture.timer->CCER = tmpccer;
+#ifdef PIOS_VIDEO_INPUT_FILTER
+	if (cfg->hsync_capture.timer_chan == TIM_Channel_1) {
```
Would appreciate a small comment explaining what IC1F is/does so people don't need to refer to the STM32 ref. manual. :)

tracernz commented on flight/PiOS/STM32F4xx/pios_video.c:
```diff
@@ -367,15 +367,34 @@ void PIOS_Video_Init(const struct pios_video_cfg *cfg)
+		tmpccer |= 6 << 4;
+		/* 6 = Fdts/4, N=6.  APB1=42MHz, so 6 cycles at 10.5Mhz...
+		 * require a steady value for half a microsecond
+		 */
```
This is good, but doesn't explain what/how Fdts or N mean/are derived.

mlyle committed:
pios_video: attempt to document, fix value used

mlyle commented:
jenkins, build artifacts please

mlyle commented:
Have confirmed that on this version, the OSD works.  @litterbug-- can you see if these artifacts fix the crash scenario?

It looks like to me if csync toggled 65535 times during a frame before, we'd have entered an infinite loop.  Now there's multiple lines of defense against that (input filtering to not respond to glitches; and the actual bug in the interrupt handler is fixed).

@hshred -- do you want this fix for the original brain?  If so, can you please certify it on brain hw?  (Instead, the line handler removal PR that I have open that I've asked you to look at repeatedly would fix this issue, too, if you'd prefer to go down that route...)

mlyle commented:
LOL, correction to names.  @mluessi, @litterbugs, see above please

LitterBugs commented:
I'll give it a test tonight.  :+1:

mluessi commented:
@mlyle I can test it on brain if it fixes the problem on seppuku.

By "PR that I have open that I've asked you to look at repeatedly" do you mean #1523? I had tested in on brain and it wasn't working, so I'm not sure what you mean.

Edit: Just noticed you actually made changes to #1523. I can test it again if you think it's better and it also fixes this problem.

mlyle commented:
@mluessi Yes, I've asked you about #1523 four (now five) times since the changes.  It works on Seppuku/PAL.  You failed before on Brain/NTSC.  I need to get out NTSC stuff to test still.

mlyle committed:
pios_video: further changes

1. Manage active_line differently-- and better skip duplicate vsyncs.
2. Ensure that pixel timer trigger and SPIs are disabled during vsync.
I'm concerned that this could get badly out of sync otherwise.
3. Update '2811 during at beginning of frame instead of end.

mlyle commented:
jenkins, build artifacts please

mlyle committed:
Merge remote-tracking branch 'pr/1523' into mpl-vidhax

mlyle committed:
pios_video: corrections from merge

mlyle commented:
jenkins, build artifacts please

mlyle commented:
#1523 is now a part of this.  A timer is removed and line tracking is now much simpler / easier to analyze on OSD targets.  Confirmed functional on seppuku/PAL

tracernz reviewed (approved):
This looks like a nice improvement. :+1: Ready for merge after some testing. I would suggest combinations that cover all of these at least once:
- [ ] Brain OG
- [x] Seppuku
- [x] NTSC
- [x] PAL

Also, might be nice to fix up `graphics_hight_real` in pios_video_quadspi.c too. :wink:

mlyle commented:
Seppuku PAL was tested by me; NTSC by both pug & litterbug

mlyle commented:
@mluessi ping

dustin commented:
I had a few bits of weirdness Monday with my OG Brain in my caipirinha.  It settled down after a few minutes, but was making me just a bit nervous over the ocean.

Here are several frames demonstrating weirdness.  You can also download [all the frames](https://drive.google.com/open?id=0B-RsldXxKyjhTDlfNHR3SlhEN3c) if you want to see more.

![image](https://cloud.githubusercontent.com/assets/1779/24224477/92d0b090-0f18-11e7-9ad4-ef755238b863.png)
![image](https://cloud.githubusercontent.com/assets/1779/24224483/9f8f8e00-0f18-11e7-890a-b071153884dd.png)
![image](https://cloud.githubusercontent.com/assets/1779/24224502/b34f8bfc-0f18-11e7-9897-896ea090b0e3.png)
![image](https://cloud.githubusercontent.com/assets/1779/24224532/dc331912-0f18-11e7-9eae-319973eef842.png)
![image](https://cloud.githubusercontent.com/assets/1779/24224541/ea031dee-0f18-11e7-8dab-4ebdcc04aba2.png)
![image](https://cloud.githubusercontent.com/assets/1779/24224629/500a592c-0f19-11e7-8b0a-d05c17f25d02.png)

mlyle commented:
"SENSORS" and sync anomalies sure sounds like a cousin to this problem.

tracernz commented:
@mluessi ping, can you do Brain + NTSC test?

LitterBugs commented:
@mluessi any reserve of OG Brain FCs that could be distributed for testing?

mluessi commented:
Seems to work fine on OG Brain / NTSC. 

@dustin did you have sync problems before on this plane or only with this PR?

mlyle commented:
@mluessi that was --without-- this PR

mluessi commented:
OK, that makes sense :)

mlyle commented:
@mluessi -- thanks for testing.  are you cool with this merging, then?

mluessi commented:
Yes, looks good. Thanks for the fixes
